// @mui/x-data-grid-generator/legacy/renderer/renderIncoterm.js
import * as React from 'react';
import Tooltip from '@mui/material/Tooltip';
import Box from '@mui/material/Box';
import InfoIcon from '@mui/icons-material/Info';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
var Incoterm = /*#__PURE__*/React.memo(function Incoterm(props) {
  var value = props.value;

  if (!value) {
    return null;
  }

  var valueStr = value.toString();
  var tooltip = valueStr.slice(valueStr.indexOf('(') + 1, valueStr.indexOf(')'));
  var code = valueStr.slice(0, valueStr.indexOf('(')).trim();
  return /*#__PURE__*/_jsxs(Box, {
    sx: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-between'
    },
    children: [/*#__PURE__*/_jsx("span", {
      children: code
    }), /*#__PURE__*/_jsx(Tooltip, {
      title: tooltip,
      children: /*#__PURE__*/_jsx(InfoIcon, {
        sx: {
          color: '#2196f3',
          alignSelf: 'center',
          ml: '8px'
        }
      })
    })]
  });
});
export function renderIncoterm(params) {
  return /*#__PURE__*/_jsx(Incoterm, {
    value: params.value
  });
}
