// @mui/x-data-grid-generator/node/services/random-generator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.randomUserName = exports.randomUrl = exports.randomUpdatedDate = exports.randomUnitPriceCurrency = exports.randomUnitPrice = exports.randomTraderName = exports.randomTradeDate = exports.randomTaxCode = exports.randomStatusOptions = exports.randomRating = exports.randomRateType = exports.randomRate = exports.randomQuantity = exports.randomPrice = exports.randomPnL = exports.randomPhoneNumber = exports.randomName = exports.randomMaturityDate = exports.randomJobTitle = exports.randomInt = exports.randomIncoterm = exports.randomId = exports.randomFeeRate = exports.randomEmail = exports.randomDesk = exports.randomDate = exports.randomCurrency = exports.randomCreatedDate = exports.randomCountry = exports.randomContractType = exports.randomCompanyName = exports.randomCommodity = exports.randomColor = exports.randomCity = exports.randomBrokerId = exports.randomBoolean = exports.randomArrayItem = exports.randomAddress = exports.random = exports.generateIsFilled = exports.generateFilledQuantity = void 0;

var _chance = _interopRequireDefault(require("chance"));

var _staticData = require("./static-data");

const chanceId = (0, _chance.default)();
let chance;

if (process.env.DISABLE_CHANCE_RANDOM) {
  chance = (0, _chance.default)(() => 0.5);
} else {
  chance = chanceId;
}

/**
 * Wrap a data generator that returns a string and add a prefix if the value generated has already been given
 */
const uniquenessHandler = generator => (data, context) => {
  var _context$values$rawVa;

  const rawValue = generator(data, context);

  if (!context.values) {
    return rawValue;
  }

  const valueCount = ((_context$values$rawVa = context.values[rawValue]) != null ? _context$values$rawVa : 0) + 1;
  context.values[rawValue] = valueCount + 1;

  if (valueCount > 1) {
    return `${rawValue} ${valueCount}`;
  }

  return rawValue;
};

function dateFuture(years, refDate) {
  let date = new Date();

  if (typeof refDate !== 'undefined') {
    date = new Date(Date.parse(refDate));
  }

  const range = {
    min: 1000,
    max: (years || 1) * 365 * 24 * 3600 * 1000
  }; // some time from now to N years later, in milliseconds

  const past = date.getTime() + chance.integer(range);
  date.setTime(past);
  return date;
}

function dateRecent(days, refDate) {
  let date = new Date();

  if (typeof refDate !== 'undefined') {
    date = new Date(Date.parse(refDate));
  }

  const range = {
    min: 1000,
    max: (days || 1) * 24 * 3600 * 1000
  }; // some time from now to N days ago, in milliseconds

  const past = date.getTime() - chance.integer(range);
  date.setTime(past);
  return date;
}

function datePast(years, refDate) {
  let date = new Date();

  if (typeof refDate !== 'undefined') {
    date = new Date(Date.parse(refDate));
  }

  const range = {
    min: 1000,
    max: (years || 1) * 365 * 24 * 3600 * 1000
  }; // some time from now to N years ago, in milliseconds

  const past = date.getTime() - chance.integer(range);
  date.setTime(past);
  return date;
}

const random = (min, max) => chance.floating({
  min,
  max
});

exports.random = random;

const randomInt = (min, max) => Math.floor(random(0, 1) * (max - min + 1) + min);

exports.randomInt = randomInt;

const randomPrice = (min = 0, max = 100000) => Number(random(min, max).toFixed(2));

exports.randomPrice = randomPrice;

const randomRate = () => random(0, 1);

exports.randomRate = randomRate;

const randomDate = (start, end) => new Date(start.getTime() + chance.floating({
  min: 0,
  max: 1
}) * (end.getTime() - start.getTime()));

exports.randomDate = randomDate;

const randomArrayItem = arr => arr[randomInt(0, arr.length - 1)];

exports.randomArrayItem = randomArrayItem;

const randomBoolean = () => randomArrayItem([true, false]);

exports.randomBoolean = randomBoolean;

const randomColor = () => randomArrayItem(_staticData.COLORS);

exports.randomColor = randomColor;

const randomId = () => chanceId.guid();

exports.randomId = randomId;

const randomDesk = () => `D-${chance.integer({
  min: 0,
  max: 10000
})}`;

exports.randomDesk = randomDesk;

const randomCommodity = () => randomArrayItem(_staticData.COMMODITY_OPTIONS);

exports.randomCommodity = randomCommodity;

const randomTraderName = () => chance.name();

exports.randomTraderName = randomTraderName;

const randomUserName = () => chance.twitter();

exports.randomUserName = randomUserName;

const randomEmail = () => chance.email();

exports.randomEmail = randomEmail;

const randomUrl = () => chance.url();

exports.randomUrl = randomUrl;

const randomPhoneNumber = () => chance.phone();

exports.randomPhoneNumber = randomPhoneNumber;

const randomUnitPrice = () => randomPrice(1, 100);

exports.randomUnitPrice = randomUnitPrice;

const randomUnitPriceCurrency = () => randomArrayItem(_staticData.CURRENCY_OPTIONS);

exports.randomUnitPriceCurrency = randomUnitPriceCurrency;

const randomQuantity = () => randomInt(1000, 100000);

exports.randomQuantity = randomQuantity;

const randomFeeRate = () => Number(random(0.1, 0.4).toFixed(3));

exports.randomFeeRate = randomFeeRate;

const randomIncoterm = () => randomArrayItem(_staticData.INCOTERM_OPTIONS);

exports.randomIncoterm = randomIncoterm;

const randomStatusOptions = () => randomArrayItem(_staticData.STATUS_OPTIONS);

exports.randomStatusOptions = randomStatusOptions;

const randomPnL = () => random(-100000000, 100000000);

exports.randomPnL = randomPnL;

const randomMaturityDate = () => dateFuture();

exports.randomMaturityDate = randomMaturityDate;

const randomTradeDate = () => dateRecent();

exports.randomTradeDate = randomTradeDate;

const randomBrokerId = () => chance.guid();

exports.randomBrokerId = randomBrokerId;

const randomCompanyName = () => chance.company();

exports.randomCompanyName = randomCompanyName;

const randomCountry = () => randomArrayItem(_staticData.COUNTRY_ISO_OPTIONS);

exports.randomCountry = randomCountry;

const randomCurrency = () => randomArrayItem(_staticData.CURRENCY_OPTIONS);

exports.randomCurrency = randomCurrency;

const randomAddress = () => chance.address();

exports.randomAddress = randomAddress;

const randomCity = () => chance.city();

exports.randomCity = randomCity;

const randomTaxCode = () => randomArrayItem(_staticData.TAXCODE_OPTIONS);

exports.randomTaxCode = randomTaxCode;

const randomContractType = () => randomArrayItem(_staticData.CONTRACT_TYPE_OPTIONS);

exports.randomContractType = randomContractType;

const randomRateType = () => randomArrayItem(_staticData.RATE_TYPE_OPTIONS);

exports.randomRateType = randomRateType;

const randomCreatedDate = () => datePast();

exports.randomCreatedDate = randomCreatedDate;

const randomUpdatedDate = () => dateRecent();

exports.randomUpdatedDate = randomUpdatedDate;

const randomJobTitle = () => chance.profession();

exports.randomJobTitle = randomJobTitle;

const randomRating = () => randomInt(1, 5);

exports.randomRating = randomRating;
const randomName = uniquenessHandler(() => chance.name());
exports.randomName = randomName;

const generateFilledQuantity = data => Number((data.quantity * randomRate()).toFixed()) / data.quantity;

exports.generateFilledQuantity = generateFilledQuantity;

const generateIsFilled = data => data.quantity === data.filledQuantity;

exports.generateIsFilled = generateIsFilled;
