// @mui/x-data-grid-generator/node/services/tree-data-generator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addTreeDataOptionsToDemoData = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _randomGenerator = require("./random-generator");

const addTreeDataOptionsToDemoData = (data, options = {}) => {
  var _groupingCol$headerNa;

  const {
    averageChildren = 2,
    maxDepth = 1,
    groupingField
  } = options;
  const hasTreeData = maxDepth > 1 && groupingField != null;

  if (!hasTreeData) {
    return data;
  }

  if (data.rows.length > 1000) {
    throw new Error('MUI: useDemoData tree data mode only works up to 1000 rows.');
  }

  const rowsByTreeDepth = {};
  const rowsCount = data.rows.length;
  const groupingCol = data.columns.find(col => col.field === options.groupingField);

  if (!groupingCol) {
    throw new Error('MUI: The tree data grouping field does not exist');
  }

  data.initialState.columns.columnVisibilityModel[groupingField] = false;

  for (let i = 0; i < rowsCount; i += 1) {
    const row = data.rows[i];
    const currentChunk = Math.floor(i * (averageChildren ** maxDepth - 1) / rowsCount) + 1;
    const currentDepth = Math.floor(Math.log(currentChunk) / Math.log(averageChildren));

    if (!rowsByTreeDepth[currentDepth]) {
      rowsByTreeDepth[currentDepth] = {
        rows: {},
        rowIndexes: []
      };
    }

    rowsByTreeDepth[currentDepth].rows[i] = {
      value: row,
      parentIndex: null
    };
    rowsByTreeDepth[currentDepth].rowIndexes.push(i);
  }

  Object.entries(rowsByTreeDepth).forEach(([depthStr, {
    rows
  }]) => {
    const depth = Number(depthStr);
    Object.values(rows).forEach(row => {
      const path = [];
      let previousRow = null;

      for (let k = depth; k >= 0; k -= 1) {
        let rowTemp;

        if (k === depth) {
          if (depth > 0) {
            row.parentIndex = Number((0, _randomGenerator.randomArrayItem)(rowsByTreeDepth[depth - 1].rowIndexes));
          }

          rowTemp = row;
        } else {
          rowTemp = rowsByTreeDepth[k].rows[previousRow.parentIndex];
        }

        path.unshift(rowTemp.value[groupingField]);
        previousRow = rowTemp;
      }

      row.value.path = path;
    });
  });
  return (0, _extends2.default)({}, data, {
    groupingColDef: {
      headerName: (_groupingCol$headerNa = groupingCol.headerName) != null ? _groupingCol$headerNa : groupingCol.field,
      width: 250
    },
    getTreeDataPath: row => row.path,
    treeData: true
  });
};

exports.addTreeDataOptionsToDemoData = addTreeDataOptionsToDemoData;
