// @mui/x-data-grid-generator/legacy/hooks/useDemoData.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _typeof from "@babel/runtime/helpers/esm/typeof";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

import * as React from 'react';
import LRUCache from 'lru-cache';
import { getRealGridData } from '../services/real-data-service';
import { getCommodityColumns } from '../columns/commodities.columns';
import { getEmployeeColumns } from '../columns/employees.columns';
import asyncWorker from '../services/asyncWorker';
import { addTreeDataOptionsToDemoData } from '../services/tree-data-generator';
var dataCache = new LRUCache({
  max: 10,
  ttl: 60 * 5 * 1e3 // 5 minutes

});

// Generate fake data from a seed.
// It's about x20 faster than getRealData.
function extrapolateSeed(_x, _x2) {
  return _extrapolateSeed.apply(this, arguments);
}

function _extrapolateSeed() {
  _extrapolateSeed = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(rowLength, data) {
    return _regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            return _context2.abrupt("return", new Promise(function (resolve) {
              var seed = data.rows;
              var rows = data.rows.slice();
              var tasks = {
                current: rowLength - seed.length
              };

              function work() {
                var row = {};

                for (var j = 0; j < data.columns.length; j += 1) {
                  var column = data.columns[j];
                  var index = Math.round(Math.random() * (seed.length - 1));

                  if (column.field === 'id') {
                    row.id = "id-".concat(tasks.current + seed.length);
                  } else {
                    row[column.field] = seed[index][column.field];
                  }
                }

                rows.push(row);
                tasks.current -= 1;
              }

              asyncWorker({
                work: work,
                done: function done() {
                  return resolve(_extends({}, data, {
                    rows: rows
                  }));
                },
                tasks: tasks
              });
            }));

          case 1:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _extrapolateSeed.apply(this, arguments);
}

var deepFreeze = function deepFreeze(object) {
  // Retrieve the property names defined on object
  var propNames = Object.getOwnPropertyNames(object); // Freeze properties before freezing self
  // eslint-disable-next-line no-restricted-syntax

  var _iterator = _createForOfIteratorHelper(propNames),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var name = _step.value;
      var value = object[name];

      if (value && _typeof(value) === 'object') {
        deepFreeze(value);
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return Object.freeze(object);
};

export var useDemoData = function useDemoData(options) {
  var _options$treeData4, _options$treeData5, _options$treeData6;

  var _React$useState = React.useState(options.rowLength),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      rowLength = _React$useState2[0],
      setRowLength = _React$useState2[1];

  var _React$useState3 = React.useState(0),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      index = _React$useState4[0],
      setIndex = _React$useState4[1];

  var _React$useState5 = React.useState(true),
      _React$useState6 = _slicedToArray(_React$useState5, 2),
      loading = _React$useState6[0],
      setLoading = _React$useState6[1];

  var getColumns = React.useCallback(function () {
    var columns = options.dataSet === 'Commodity' ? getCommodityColumns(options.editable) : getEmployeeColumns();

    if (options.visibleFields) {
      columns = columns.map(function (col) {
        var _options$visibleField;

        return (_options$visibleField = options.visibleFields) != null && _options$visibleField.includes(col.field) ? col : _extends({}, col, {
          hide: true
        });
      });
    }

    if (options.maxColumns) {
      columns = columns.slice(0, options.maxColumns);
    }

    return columns;
  }, [options.dataSet, options.editable, options.maxColumns, options.visibleFields]);

  var _React$useState7 = React.useState(function () {
    var columns = getColumns(); // TODO v6: Stop using `GridColDef.hide`

    // TODO v6: Stop using `GridColDef.hide`
    var columnVisibilityModel = {};
    columns.forEach(function (col) {
      if (col.hide) {
        columnVisibilityModel[col.field] = false;
      }
    });
    return addTreeDataOptionsToDemoData({
      columns: columns,
      rows: [],
      initialState: {
        columns: {
          columnVisibilityModel: columnVisibilityModel
        }
      }
    }, options.treeData);
  }),
      _React$useState8 = _slicedToArray(_React$useState7, 2),
      data = _React$useState8[0],
      setData = _React$useState8[1];

  React.useEffect(function () {
    var cacheKey = "".concat(options.dataSet, "-").concat(rowLength, "-").concat(index, "-").concat(options.maxColumns); // Cache to allow fast switch between the JavaScript and TypeScript version
    // of the demos.

    if (dataCache.has(cacheKey)) {
      var newData = dataCache.get(cacheKey);
      setData(newData);
      setLoading(false);
      return undefined;
    }

    var active = true;

    _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var _options$treeData, _options$treeData2, _options$treeData3;

      var newData;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              setLoading(true);

              if (!(rowLength > 1000)) {
                _context.next = 10;
                break;
              }

              _context.next = 4;
              return getRealGridData(1000, getColumns());

            case 4:
              newData = _context.sent;
              _context.next = 7;
              return extrapolateSeed(rowLength, newData);

            case 7:
              newData = _context.sent;
              _context.next = 13;
              break;

            case 10:
              _context.next = 12;
              return getRealGridData(rowLength, getColumns());

            case 12:
              newData = _context.sent;

            case 13:
              if (active) {
                _context.next = 15;
                break;
              }

              return _context.abrupt("return");

            case 15:
              newData = addTreeDataOptionsToDemoData(newData, {
                maxDepth: (_options$treeData = options.treeData) == null ? void 0 : _options$treeData.maxDepth,
                groupingField: (_options$treeData2 = options.treeData) == null ? void 0 : _options$treeData2.groupingField,
                averageChildren: (_options$treeData3 = options.treeData) == null ? void 0 : _options$treeData3.averageChildren
              }); // It's quite slow. No need for it in production.

              if (process.env.NODE_ENV !== 'production') {
                deepFreeze(newData);
              }

              dataCache.set(cacheKey, newData);
              setData(newData);
              setLoading(false);

            case 20:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }))();

    return function () {
      active = false;
    };
  }, [rowLength, options.dataSet, options.maxColumns, (_options$treeData4 = options.treeData) == null ? void 0 : _options$treeData4.maxDepth, (_options$treeData5 = options.treeData) == null ? void 0 : _options$treeData5.groupingField, (_options$treeData6 = options.treeData) == null ? void 0 : _options$treeData6.averageChildren, index, getColumns]);
  return {
    data: data,
    loading: loading,
    setRowLength: setRowLength,
    loadNewData: function loadNewData() {
      setIndex(function (oldIndex) {
        return oldIndex + 1;
      });
    }
  };
};
