// @mui/x-data-grid-generator/legacy/services/random-generator.js
import globalChance from 'chance';
import { COLORS, COMMODITY_OPTIONS, CONTRACT_TYPE_OPTIONS, COUNTRY_ISO_OPTIONS, CURRENCY_OPTIONS, INCOTERM_OPTIONS, RATE_TYPE_OPTIONS, STATUS_OPTIONS, TAXCODE_OPTIONS } from './static-data';
var chanceId = globalChance();
var chance;

if (process.env.DISABLE_CHANCE_RANDOM) {
  chance = globalChance(function () {
    return 0.5;
  });
} else {
  chance = chanceId;
}

/**
 * Wrap a data generator that returns a string and add a prefix if the value generated has already been given
 */
var uniquenessHandler = function uniquenessHandler(generator) {
  return function (data, context) {
    var _context$values$rawVa;

    var rawValue = generator(data, context);

    if (!context.values) {
      return rawValue;
    }

    var valueCount = ((_context$values$rawVa = context.values[rawValue]) != null ? _context$values$rawVa : 0) + 1;
    context.values[rawValue] = valueCount + 1;

    if (valueCount > 1) {
      return "".concat(rawValue, " ").concat(valueCount);
    }

    return rawValue;
  };
};

function dateFuture(years, refDate) {
  var date = new Date();

  if (typeof refDate !== 'undefined') {
    date = new Date(Date.parse(refDate));
  }

  var range = {
    min: 1000,
    max: (years || 1) * 365 * 24 * 3600 * 1000
  }; // some time from now to N years later, in milliseconds

  var past = date.getTime() + chance.integer(range);
  date.setTime(past);
  return date;
}

function dateRecent(days, refDate) {
  var date = new Date();

  if (typeof refDate !== 'undefined') {
    date = new Date(Date.parse(refDate));
  }

  var range = {
    min: 1000,
    max: (days || 1) * 24 * 3600 * 1000
  }; // some time from now to N days ago, in milliseconds

  var past = date.getTime() - chance.integer(range);
  date.setTime(past);
  return date;
}

function datePast(years, refDate) {
  var date = new Date();

  if (typeof refDate !== 'undefined') {
    date = new Date(Date.parse(refDate));
  }

  var range = {
    min: 1000,
    max: (years || 1) * 365 * 24 * 3600 * 1000
  }; // some time from now to N years ago, in milliseconds

  var past = date.getTime() - chance.integer(range);
  date.setTime(past);
  return date;
}

export var random = function random(min, max) {
  return chance.floating({
    min: min,
    max: max
  });
};
export var randomInt = function randomInt(min, max) {
  return Math.floor(random(0, 1) * (max - min + 1) + min);
};
export var randomPrice = function randomPrice() {
  var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 100000;
  return Number(random(min, max).toFixed(2));
};
export var randomRate = function randomRate() {
  return random(0, 1);
};
export var randomDate = function randomDate(start, end) {
  return new Date(start.getTime() + chance.floating({
    min: 0,
    max: 1
  }) * (end.getTime() - start.getTime()));
};
export var randomArrayItem = function randomArrayItem(arr) {
  return arr[randomInt(0, arr.length - 1)];
};
export var randomBoolean = function randomBoolean() {
  return randomArrayItem([true, false]);
};
export var randomColor = function randomColor() {
  return randomArrayItem(COLORS);
};
export var randomId = function randomId() {
  return chanceId.guid();
};
export var randomDesk = function randomDesk() {
  return "D-".concat(chance.integer({
    min: 0,
    max: 10000
  }));
};
export var randomCommodity = function randomCommodity() {
  return randomArrayItem(COMMODITY_OPTIONS);
};
export var randomTraderName = function randomTraderName() {
  return chance.name();
};
export var randomUserName = function randomUserName() {
  return chance.twitter();
};
export var randomEmail = function randomEmail() {
  return chance.email();
};
export var randomUrl = function randomUrl() {
  return chance.url();
};
export var randomPhoneNumber = function randomPhoneNumber() {
  return chance.phone();
};
export var randomUnitPrice = function randomUnitPrice() {
  return randomPrice(1, 100);
};
export var randomUnitPriceCurrency = function randomUnitPriceCurrency() {
  return randomArrayItem(CURRENCY_OPTIONS);
};
export var randomQuantity = function randomQuantity() {
  return randomInt(1000, 100000);
};
export var randomFeeRate = function randomFeeRate() {
  return Number(random(0.1, 0.4).toFixed(3));
};
export var randomIncoterm = function randomIncoterm() {
  return randomArrayItem(INCOTERM_OPTIONS);
};
export var randomStatusOptions = function randomStatusOptions() {
  return randomArrayItem(STATUS_OPTIONS);
};
export var randomPnL = function randomPnL() {
  return random(-100000000, 100000000);
};
export var randomMaturityDate = function randomMaturityDate() {
  return dateFuture();
};
export var randomTradeDate = function randomTradeDate() {
  return dateRecent();
};
export var randomBrokerId = function randomBrokerId() {
  return chance.guid();
};
export var randomCompanyName = function randomCompanyName() {
  return chance.company();
};
export var randomCountry = function randomCountry() {
  return randomArrayItem(COUNTRY_ISO_OPTIONS);
};
export var randomCurrency = function randomCurrency() {
  return randomArrayItem(CURRENCY_OPTIONS);
};
export var randomAddress = function randomAddress() {
  return chance.address();
};
export var randomCity = function randomCity() {
  return chance.city();
};
export var randomTaxCode = function randomTaxCode() {
  return randomArrayItem(TAXCODE_OPTIONS);
};
export var randomContractType = function randomContractType() {
  return randomArrayItem(CONTRACT_TYPE_OPTIONS);
};
export var randomRateType = function randomRateType() {
  return randomArrayItem(RATE_TYPE_OPTIONS);
};
export var randomCreatedDate = function randomCreatedDate() {
  return datePast();
};
export var randomUpdatedDate = function randomUpdatedDate() {
  return dateRecent();
};
export var randomJobTitle = function randomJobTitle() {
  return chance.profession();
};
export var randomRating = function randomRating() {
  return randomInt(1, 5);
};
export var randomName = uniquenessHandler(function () {
  return chance.name();
});
export var generateFilledQuantity = function generateFilledQuantity(data) {
  return Number((data.quantity * randomRate()).toFixed()) / data.quantity;
};
export var generateIsFilled = function generateIsFilled(data) {
  return data.quantity === data.filledQuantity;
};
