// @mui/x-data-grid-generator/legacy/services/tree-data-generator.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { randomArrayItem } from './random-generator';
export var addTreeDataOptionsToDemoData = function addTreeDataOptionsToDemoData(data) {
  var _groupingCol$headerNa;

  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var _options$averageChild = options.averageChildren,
      averageChildren = _options$averageChild === void 0 ? 2 : _options$averageChild,
      _options$maxDepth = options.maxDepth,
      maxDepth = _options$maxDepth === void 0 ? 1 : _options$maxDepth,
      groupingField = options.groupingField;
  var hasTreeData = maxDepth > 1 && groupingField != null;

  if (!hasTreeData) {
    return data;
  }

  if (data.rows.length > 1000) {
    throw new Error('MUI: useDemoData tree data mode only works up to 1000 rows.');
  }

  var rowsByTreeDepth = {};
  var rowsCount = data.rows.length;
  var groupingCol = data.columns.find(function (col) {
    return col.field === options.groupingField;
  });

  if (!groupingCol) {
    throw new Error('MUI: The tree data grouping field does not exist');
  }

  data.initialState.columns.columnVisibilityModel[groupingField] = false;

  for (var i = 0; i < rowsCount; i += 1) {
    var row = data.rows[i];
    var currentChunk = Math.floor(i * (Math.pow(averageChildren, maxDepth) - 1) / rowsCount) + 1;
    var currentDepth = Math.floor(Math.log(currentChunk) / Math.log(averageChildren));

    if (!rowsByTreeDepth[currentDepth]) {
      rowsByTreeDepth[currentDepth] = {
        rows: {},
        rowIndexes: []
      };
    }

    rowsByTreeDepth[currentDepth].rows[i] = {
      value: row,
      parentIndex: null
    };
    rowsByTreeDepth[currentDepth].rowIndexes.push(i);
  }

  Object.entries(rowsByTreeDepth).forEach(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2),
        depthStr = _ref2[0],
        rows = _ref2[1].rows;

    var depth = Number(depthStr);
    Object.values(rows).forEach(function (row) {
      var path = [];
      var previousRow = null;

      for (var k = depth; k >= 0; k -= 1) {
        var rowTemp = void 0;

        if (k === depth) {
          if (depth > 0) {
            row.parentIndex = Number(randomArrayItem(rowsByTreeDepth[depth - 1].rowIndexes));
          }

          rowTemp = row;
        } else {
          rowTemp = rowsByTreeDepth[k].rows[previousRow.parentIndex];
        }

        path.unshift(rowTemp.value[groupingField]);
        previousRow = rowTemp;
      }

      row.value.path = path;
    });
  });
  return _extends({}, data, {
    groupingColDef: {
      headerName: (_groupingCol$headerNa = groupingCol.headerName) != null ? _groupingCol$headerNa : groupingCol.field,
      width: 250
    },
    getTreeDataPath: function getTreeDataPath(row) {
      return row.path;
    },
    treeData: true
  });
};
